### PR TITLE
add configuration for running local shellcheck script

### PR DESCRIPTION
**Pull request by cnygardtw**
- avoid running via docker image, use (kotlin) `useDocker.set(false)`
- configure path to shellcheck binary via shellcheckBinary property
- add tests for all current docker-based tests
- clean up config property definitions, add config for image, image version
- upgrade to gradle v7.0
- update docs
- bump version to 1.1.0

### CODE REVIEW

**Comment by felipefzdz** [comment]
Thanks a lot for the PR! I'll review it asap.

File: README.adoc [hunk 1]
```diff
@@ -36,21 +36,41 @@ The Shellcheck plugin a task called `shellcheck` to the project.
...
+}
+----
+
+[source,groovy]
+----
 shellcheck {
     sources = files("src/shellScripts", "src/moreScripts")
-    isIgnoreFailures = true
-    isShowViolations = true
+    continueBuildOnFailure = true
```
Comment (felipefzdz) [inline README.adoc:56]: `continueBuildOnFailure` and `showViolations` are breaking changes so they shouldn't be part of a minor release, like 1.1. Even if we could argue that they're an improvement from the previous names, I don't think that they should require a major release. `continueBuildOnFailure` refactor was also incomplete, as there are some `ignoreFailures` references still around. Would you mind reverting this refactor?

File: README.adoc [hunk 2]
```diff
@@ -36,21 +36,41 @@ The Shellcheck plugin a task called `shellcheck` to the project.
...
+    showViolations = true
+    useDocker = true
+    shellcheckImage = "koalaman/shellcheck-alpine"
```
Comment (felipefzdz) [inline README.adoc:59]: Why is `shellcheckImage` now configurable? i.e. is it any valuable scenario where we see a user providing something like `pipelinecomponents/shellcheck`? I'm not familiar with the inner details of the different docker images wrapping Shellcheck, but if there's no known value, I'd refrain adding this configuration.

File: README.adoc [hunk 3]
```diff
@@ -36,21 +36,41 @@ The Shellcheck plugin a task called `shellcheck` to the project.
...
 ----
 
 * sources - Folders where the shell scripts are located. It will search recursively matching `.sh`, `.bash`,
 `.bash_login`, `.bash_logout`, `.bash_profile`, `.bashrc` and `.ksh` files. Matching on shell scripts with no extension
 is not supported yet.
-* isIgnoreFailures - Whether to allow the build to continue if there are warnings.
-* isShowViolations - Whether rule violations are to be displayed on the console. Defaults to true.
+* continueBuildOnFailure - Whether to allow the build to continue if there are warnings.
+* showViolations - Whether rule violations are to be displayed on the console. Defaults to true.
+* useDocker - Whether to use docker image (true) or local shellcheck binary (false)
```
Comment (felipefzdz) [inline README.adoc:71]: Could you please add the default here? true, in this case.

File: shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    def "(local binary) can clear configuration cache"() {
+        given:
+
+        when:
+        clearConfigurationCache()
+
+        then:
+        //localGradleDirectory.exists()
```
Comment (felipefzdz) [inline shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy:56]: Why this commented line?

File: shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        !localConfigCacheDirectory.exists()
+    }
+
+
+    def "(local binary) shellcheck task can be loaded from the configuration cache"() {
+        given:
+        buildFile << """
+shellcheck {
+    sources.set(files("${resources.absolutePath}/without_violations"))
+    useDocker.set(false)
```
Comment (felipefzdz) [inline shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy:66]: (local binary)  tests, i.e. the ones that contain this line on the to be tested build.file are not reproducible anymore, in the sense that they require `/usr/local/bin/shellcheck` to be installed on the machine executing the tests, being a local machine or a CI agent.

I don't think that we should get too smart here, trying to install it if not available, but a message stating the reason of this to fail would be appreciated. Therefore, I'd organize the tests in two classes at least, the ones that are portable (Docker based ones) and the local that are not. On the latter we could add a single comment opening the class, or even an `@IgnoreIf` that would check for `/usr/local/bin/shellcheck` existence. Then if we decide to add CI support for this project we won't be blocked by these failing tests.

File: shellcheck/src/main/java/com/felipefzdz/gradle/shellcheck/Shellcheck.java [hunk 6]
```diff
@@ -23,39 +29,64 @@
...
+     */
+    @Console
+    abstract public Property<Boolean> getShowViolations();
+
+    /**
+     * Configure whether shellcheck is run via Docker container or local command line.
+     * @return the flag value, true == docker
+     */
+    @Input
+//    @Optional
```
Comment (felipefzdz) [inline shellcheck/src/main/java/com/felipefzdz/gradle/shellcheck/Shellcheck.java:52]: Why these `@Optional` commented out?

**Review by felipefzdz — CHANGES_REQUESTED** [review]
Thanks for tackling this! Left some comments.

**Comment by felipefzdz** [comment]
@cnygardtw I'll merge this and do the suggested changes myself on the `main` branch. I'll ping you when this is released. Thanks again for the PR :)

**Comment by felipefzdz** [comment]
@cnygardtw just to let you know that this has been released, along with the ability of self install the binary.

Thanks!

**Comment by cnygardtw** [comment]
Thanks, appreciate you picking this up.  Sorry I didn't have time at the moment to push the updates.   I'll check it out.

File: README.adoc (hunk 1 above)
Comment (cnygardtw) [reply README.adoc:56]: Just to follow up, I was following the latest guidance from Gradle re: deprecated Convention mapping, but the parent class still used it.  So I had to rename the property to work around that.  I was not trying to improve the names or break compatibility...

File: README.adoc (hunk 2 above)
Comment (cnygardtw) [reply README.adoc:59]: If an org has restrictions on using random docker images from the web, they could configure an internally-created shellcheck image instead.  I think it has value, and providing a default does no harm.

File: shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy (hunk 4 above)
Comment (cnygardtw) [reply shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy:56]: honestly not sure, probably cruft

File: shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy (hunk 5 above)
Comment (cnygardtw) [reply shellcheck/src/functionalTest/groovy/com/felipefzdz/gradle/shellcheck/ShellcheckPluginCacheTest.groovy:66]: Sounds reasonable. I was assuming that the docker image which does the build/test in CI would need to have shellcheck installed locally (and therefore was outside my ability to change).

File: shellcheck/src/main/java/com/felipefzdz/gradle/shellcheck/Shellcheck.java (hunk 6 above)
Comment (cnygardtw) [reply shellcheck/src/main/java/com/felipefzdz/gradle/shellcheck/Shellcheck.java:52]: Should probably have removed them.  Providing default values in the property conventions means they don't need to be optional, and supplied defaults makes the code simpler by not dealing with missing values.

File: README.adoc (hunk 1 above)
Comment (felipefzdz) [reply README.adoc:56]: You're right. I misunderstood the change. I'll tackle that in the future!

File: README.adoc (hunk 2 above)
Comment (felipefzdz) [reply README.adoc:59]: Makes sense, thanks for the explanation. I'll put it back.

**Pull request merged**